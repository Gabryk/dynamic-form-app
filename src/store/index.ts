import { configureStore } from "@reduxjs/toolkit";
import userApplicationReducer from "./userApplicationSlice";

export const reducer = { userApplication: userApplicationReducer };

export const store = configureStore({
  reducer,
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
