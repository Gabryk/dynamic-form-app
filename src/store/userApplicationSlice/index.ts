import { PayloadAction, createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import {
  RejectedActionFromAsyncThunk,
  FulfilledActionFromAsyncThunk,
  PendingActionFromAsyncThunk,
} from "@reduxjs/toolkit/dist/matchers";
import fakeApplication from "services/fakeApplication";

type Data = { [x: string]: string | number };
export interface UserApplication {
  loading?: boolean;
  error?: string;
  data?: Data;
}

const initialState: UserApplication = {};

export const uploadUserApplication = createAsyncThunk(
  "user/upload",
  async (data: Data, { signal }) => {
    const [promise, abort] = fakeApplication(data);

    signal.addEventListener("abort", () => {
      abort();
    });

    return await promise;
  }
);
const setSuccessUserApplication = (
  _: UserApplication,
  action: FulfilledActionFromAsyncThunk<typeof uploadUserApplication>
) => ({
  loading: false,
  data: action.payload,
});
const setLoadingUserApplication = (
  _: UserApplication,
  action: PendingActionFromAsyncThunk<typeof uploadUserApplication>
) => ({
  loading: true,
});
const setFailUserApplication = (
  _: UserApplication,
  action: RejectedActionFromAsyncThunk<typeof uploadUserApplication>
): UserApplication => ({
  loading: false,
  error: action.error.message,
});

const setUserData = (
  _: UserApplication,
  action: PayloadAction<Data>
): UserApplication => {
  return { data: action.payload, loading: false };
};

export const userApplicationSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    setUserData,
  },
  extraReducers(builder) {
    builder.addCase(uploadUserApplication.pending, setLoadingUserApplication);
    builder.addCase(uploadUserApplication.fulfilled, (state, action) =>
      setSuccessUserApplication(state, action)
    );
    builder.addCase(uploadUserApplication.rejected, setFailUserApplication);
  },
});

export const { setUserData: setUserDataAction } = userApplicationSlice.actions;

export default userApplicationSlice.reducer;
