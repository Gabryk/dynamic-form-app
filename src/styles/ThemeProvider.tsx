import { ThemeProvider as Provider } from "styled-components";
import theme from "styles/Theme";
import { PropsWithChildren } from "react";

const ThemeProvider = ({ children }: PropsWithChildren) => {
  return <Provider theme={theme}>{children}</Provider>;
};

export default ThemeProvider;
