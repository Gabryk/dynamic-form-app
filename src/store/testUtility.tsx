import React, { PropsWithChildren } from "react";
import { combineReducers, configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { reducer } from "store";

// Create the root reducer separately so we can extract the RootState type
const rootReducer = combineReducers(reducer);

const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

type RootState = ReturnType<typeof rootReducer>;
type AppStore = ReturnType<typeof setupStore>;
