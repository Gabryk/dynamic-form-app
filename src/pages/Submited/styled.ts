import styled from "styled-components";
import { fieldStyled } from "styles/Global";

export const FieldStyled = styled.div`
  ${fieldStyled}
`;
