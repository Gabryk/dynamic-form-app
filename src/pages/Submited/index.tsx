import { useSelector } from "react-redux";
import { RootState } from "store";
import { Field } from "constants/field-set";
import DynamicLayout from "components/DynamicLayout";
import Header from "components/Header";
import { FieldStyled } from "./styled";

const Submited = () => {
  const { data } = useSelector((state: RootState) => state.userApplication);

  const renderField = (field: Field) => {
    return (
      <FieldStyled key={field.id}>{data?.[field.id] || "N/A"}</FieldStyled>
    );
  };

  return (
    <div>
      <Header
        title="Thanks you!"
        description="Your application was successfully submitted."
      />

      <DynamicLayout renderField={renderField} />
    </div>
  );
};

export default Submited;
