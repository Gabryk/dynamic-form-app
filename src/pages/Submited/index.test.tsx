import React from "react";
import { renderWithProviders } from "store/testUtility";
import Submited from ".";
import ThemeProvider from "styles/ThemeProvider";

const preloadedState = {
  userApplication: {
    loading: false,
    data: {
      email: "somefake@gmail.com",
      address1: "1187 Twin Willow Lane",
      city: "Wilmington",
      firstName: "Rosemary",
      jobTitle: "Engineer - lead",
      lastName: "J Jennings",
      phone: "910-520-1684",
      reason:
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam",
      state: "North Carolina",
      zip: "28412",
    },
  },
};

test("Render Submitted page", async () => {
  const { asFragment } = renderWithProviders(
    <ThemeProvider>
      <Submited />
    </ThemeProvider>,
    {
      preloadedState,
    }
  );

  expect(asFragment()).toMatchSnapshot();
});
