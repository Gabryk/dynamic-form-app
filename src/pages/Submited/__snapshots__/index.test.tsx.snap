// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Submitted page 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-dmqGxT bvEMza"
    >
      <h1>
        Thanks you!
      </h1>
      <p>
        Your application was successfully submitted.
      </p>
    </div>
    <div
      class="sc-gueWre fYygbm"
    >
      <div
        class="sc-beqWul dVXNzW"
      >
        <div
          class="sc-hLsdIM gIyUnl"
        >
          Rosemary
        </div>
        <div
          class="sc-hLsdIM gIyUnl"
        >
          J Jennings
        </div>
      </div>
      <div
        class="sc-hLsdIM gIyUnl"
      >
        somefake@gmail.com
      </div>
      <div
        class="sc-hLsdIM gIyUnl"
      >
        1187 Twin Willow Lane
      </div>
      <div
        class="sc-beqWul dVXNzW"
      >
        <div
          class="sc-hLsdIM gIyUnl"
        >
          Wilmington
        </div>
        <div
          class="sc-hLsdIM gIyUnl"
        >
          North Carolina
        </div>
        <div
          class="sc-hLsdIM gIyUnl"
        >
          28412
        </div>
      </div>
      <div
        class="sc-hLsdIM gIyUnl"
      >
        910-520-1684
      </div>
      <div
        class="sc-hLsdIM gIyUnl"
      >
        Engineer - lead
      </div>
      <div
        class="sc-hLsdIM gIyUnl"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam
      </div>
    </div>
  </div>
</DocumentFragment>
`;
